### DISCUSSION TITLE
CellExplorer with Plexon data

### DISCUSSION THREAD

**Discussion by cyjzheng** [petersenpeter/CellExplorer#79] [General]
Hello, would CellExplorer work with plx data that is already high-pass filtered? We have the HLK3 system and the spike channel output is not analog raw data so the data cannot be sorted by kilosort or other sorters mentioned in the input for Cellexplorer. Is there a way we can still use CellExplorer?

**Comment by petersenpeter** [comment]
Hi 

As far as I know, you can export the plx format to a binary file. 

If this is not an option, you can create a script to import your spike data to the CellExplorer [spikes format](https://cellexplorer.org/datastructure/data-structure-and-format/#spikes). 

**Required/Recommended fields in the spikes struct:**
times: a 1xN cell-struct for N units each containing a Mx1 vector with M spike events in seconds.
UID: a 1xN vector with values 1:N.
shankID: a 1xN vector containing the corresponding shank/electrode-group each unit (1-indexed).
maxWaveformCh1: a 1xN vector with the channel for the maximum waveform for the units (1-indexed)
total: a 1xN vector with the total number of spikes for each unit.
filtWaveform: a 1xN cell-struct with spike waveforms (each a 1xP vector) from maxWaveformCh1 (µV).
numcells: number of cells.
basename: name of the session (string).
processinginfo: a substruct with information about how the spikes was generated including the name of the function, version, date and the parameters.

Once you create the spikes struct and save it to a mat file (basename.spikes.cellinfo.mat), run the processing pipeline to generate the cell metrics. You lose the standardized extraction of waveforms, but the pipeline should work.

**Comment by cyjzheng** [comment]
Hi Peter,
Thanks for the reply! I built a spikes struct as you described and saved it in the basepath as "basename.spikes.cellinfo.mat". I also made basename.session.mat file which passed the "validateSessionStruct".  I attached the two files here too.
Then I tried to run the pipeline using
>> cell_metrics = ProcessCellMetrics('basepath',basepath)  

  However, it seems like the loadSpikes.m script is still trying to load waveforms from .dat files. Seems like it couldn't load waveforms directly from the "basename.spikes.cellinfo.mat"?
Also, regarding binary files- I used the HLK3 acquisition system to collect these data. It applied a high-pass filter to the data and only output spikes and waveforms as timestamps. So even if I convert the data to a binary file, it's not continuous and I assume it wouldn't make sense to use Kilosort or other sorting tools on it.
> In ProcessCellMetrics (line 136) 
[09-01-2023 11:55:41: 20220215_T62_mix1] Loading 20220215_T62_mix1.session.mat from basepath
  Using vertical spacing from preferences
  Using layout from preferences
[09-01-2023 11:55:41: 20220215_T62_mix1] Getting spikes
loadSpikes: Using existing spikes file
Getting waveforms from dat file (nPull=600). Bandpass filter applied: 500 - 8000 Hz. No bad channels detected. 
Error using getWaveformsFromDat (line 116)
Binary file missing: D:\CE_test\20220215_T62_mix1.dat
 

[session_spikes.zip](https://github.com/petersenpeter/CellExplorer/files/10375927/session_spikes.zip)

**Comment by petersenpeter** [comment]
Hi

I downloaded your zip file and found a few things to fix:

1. The session struct was not properly defined. Please use the sessionTemplate to generate it the first time. Related, your sampling rate was set to 40,000Hz, yet you only provided waveforms with 32 samples, which seems to have been a mistake (0.8ms total waveform width)? 

2.  You were missing a few fields in the spikes struct: `.numCell` and `.timeWaveform`, `cluID`. Below are suggested values for your data:
```m
spikes.numcells = 5;
spikes.timeWaveform{5} = -0.4:1/(session.extracellular.sr/(1000)):0.39;  
% I divided by 1000 to get the sr in units of ms. Please also create timeWaveforms for the other four units.
```
3. You need to run ProcessCellMetrics with the parameter `getWaveformsFromDat` set to false, e.g.:
```m
cell_metrics = ProcessCellMetrics('getWaveformsFromDat',false);
```
Hereby it will not try to load the waveforms from a binary file.

Here is your session after my changes:
[20220215_T62_mix1.zip](https://github.com/petersenpeter/CellExplorer/files/10383271/20220215_T62_mix1.zip)

**Reply by cyjzheng** [reply]
Thank you!
the original waveform data for each unit is 128 x N (N = # of timestamps) but since it's data from tetrodes, I only picked one of the four windows with the largest trough-to-peak.
I tried to use the same session and spikes files you edited ( numcells (5), timeWaveform (1 x 5 cell), and cluID (1 x 5 double) added too) to generate cell_metrics but it doesn't pass the validateCellMetricsStruct in line 1563. Seems like the thetaModulationIndex only got generated for one unit.

cell_metrics = 
  struct with fields:
                          UID: [1 2 3 4 5]
                     ab_ratio: [0.8041 0.8296 0.8719 -1.2591 … ]
                          acg: [1×1 struct]
                acg_asymptote: [0.3962 -4.1114 -4.1114 -4.1114 … ]
                        acg_c: [4.0763 24.2454 24.2454 24.2454 … ]
                        acg_d: [6.9149 4.3704 4.3704 4.3704 4.3704]
              acg_fit_rsquare: [0.2577 NaN NaN NaN NaN]
                        acg_h: [30.8035 2.0440 2.0440 2.0440 2.0440]
                   acg_refrac: [4.2832 0.3423 0.3423 0.3423 0.3423]
                acg_tau_burst: [0.3179 1.5472 1.5472 1.5472 1.5472]
                acg_tau_decay: [11.8657 17.1774 17.1774 17.1774 … ]
                 acg_tau_rise: [0.3067 5.9666 5.9666 5.9666 5.9666]
                       animal: {'T62'  'T62'  'T62'  'T62'  'T62'}
                  brainRegion: {1×5 cell}
          burstIndex_Doublets: 3.4286
      burstIndex_Mizuseki2012: [0.0265 0.0156 0 0.1144 0.1328]
         burstIndex_Royer2012: 14.4286
                       cellID: [1 2 3 4 5]
                        cluID: [1 2 3 4 5]
                          cv2: [1.0769 1.1097 1.1564 0.9859 1.0138]
              deepSuperficial: {1×5 cell}
      deepSuperficialDistance: [NaN NaN NaN NaN NaN]
               electrodeGroup: [1 1 1 1 1]
                   firingRate: [0.1291 0.1405 0.1011 6.9988 10.1026]
                 firingRateCV: [0.4515 0.7226 0.4446 0.4460 0.3214]
          firingRateGiniCoeff: [0.2211 0.3381 0.2208 0.2253 0.1515]
                firingRateISI: [0.2487 0.4431 0.3086 14.8038 … ]
        firingRateInstability: [0.4336 0.5220 0.4041 0.1489 0.0644]
                      general: [1×1 struct]
                          isi: [1×1 struct]
                       labels: {''  ''  ''  ''  ''}
                maxWaveformCh: [1 2 1 2 3]
               maxWaveformCh1: [2 3 2 3 4]
      maxWaveformChannelOrder: [2 3 2 3 4]
                      numCell: 5
                  peakVoltage: [0.1855 0.1776 0.1111 0.1604 0.1647]
           peakVoltage_expFit: [NaN NaN NaN NaN NaN]
                     polarity: [-0.7835 -0.5682 -0.4055 0.3979 … ]
          population_modIndex: [1.0590 1.0044 0.8769 1.0459 1.0451]
             putativeCellType: {1×5 cell}
          putativeConnections: [1×1 struct]
    refractoryPeriodViolation: [0 0 0 0.2421 0.1437]
               responseCurves: [1×1 struct]
                  sessionName: {1×5 cell}
                      shankID: [1 1 1 1 1]
                   spikeCount: [530 579 417 28916 41740]
                       spikes: [1×1 struct]
        synapticConnectionsIn: [0 0 0 0 0]
       synapticConnectionsOut: [0 0 0 0 0]
         thetaModulationIndex: -0.0649
                        total: [530 579 417 28916 41740]
                     trilat_x: [NaN NaN NaN NaN NaN]
                     trilat_y: [NaN NaN NaN NaN NaN]
[session_spikes2.zip](https://github.com/petersenpeter/CellExplorer/files/10386180/session_spikes2.zip)

                 troughToPeak: [0.4417 0.4083 0.4667 0.3000 0.3000]
       troughtoPeakDerivative: [0.1333 0.1167 0.1083 0.2500 0.3083]
                    waveforms: [1×1 struct]

**Comment by petersenpeter** [comment]
Hi

I forgot to mention that I updated the repo as well. It worked with the session I shared with you, so please try again after you update your local CellExplorer installation.

```m
spikes.numcells = 5;
timeWaveform = -0.4:1/(session.extracellular.sr/(1000)):0.39;
spikes.timeWaveform{1} = timeWaveform:
spikes.timeWaveform{2} = timeWaveform:
spikes.timeWaveform{3} = timeWaveform:
spikes.timeWaveform{4} = timeWaveform:
spikes.timeWaveform{5} = timeWaveform:
```

**Reply by cyjzheng** [reply]
thank you so much! it works now.
